### PR TITLE
Feature/xml descriptions

### PR DESCRIPTION
**Pull request by DawidSA2017**
This is Story Card 1

### CODE REVIEW

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java [hunk 1]
```diff
@@ -26,6 +26,8 @@
 import java.util.LinkedList;
 import java.util.TreeSet;
 
+import javax.annotation.Nonnull;
+
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java:30]: Is this import really necessary?

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 2]
```diff
@@ -62,6 +62,7 @@
 import de.ovgu.featureide.fm.core.io.Problem;
 import de.ovgu.featureide.fm.core.io.UnsupportedModelException;
 import de.ovgu.featureide.fm.core.io.xml.XmlPropertyLoader.PropertiesParser;
+import guidsl.node;
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:65]: There should be no need to use Nodes from GUIDSL here.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 3]
```diff
@@ -296,6 +298,22 @@ protected void addDescription(Document doc, IFeature feat, Element fnod) {
 			fnod.appendChild(descr);
 		}
 	}
+	
+	protected void addDescription(Document doc, IConstraint constraint, Element fnod) {
+		// Team1 Story1
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:303]: Please avoid such comments, does not help as there is a story 1 in team 1 every year.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 4]
```diff
@@ -383,6 +401,34 @@ private void parseComments2(NodeList nodeList) throws UnsupportedModelException
...
+				String description =
+					childOfRule.getTextContent();
+
+				if ((description != null)
+					&& !description.isEmpty()) {
+					description =
+						description.replace("\t", "");
+					description =
+						description.substring(1, description.length()
+							- 1);
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:421]: This line is not clear to me.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 5]
```diff
@@ -396,19 +442,26 @@ private void parseConstraints(NodeList nodeList) throws UnsupportedModelExceptio
...
 							if (attributeName.equals(COORDINATES)) {
 								// Legacy case, for backwards compatibility
 							} else {
 								throwError("Unknown constraint attribute: " + attributeName, node);
 							}
+
 						}
 					}
+
+					// Story1
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:455]: as above

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 6]
```diff
@@ -430,6 +483,11 @@ private void parseConstraints(NodeList nodeList) throws UnsupportedModelExceptio
 				nodes.add(new Not((parseConstraints2(e.getChildNodes())).getFirst()));
 			} else if (nodeName.equals(ATMOST1)) {
 				nodes.add(new AtMost(1, parseConstraints2(e.getChildNodes())));
+			}
+				
+			else if (nodeName.equals(DESCRIPTION)) {
+		
+
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:490]: Why only empty lines here?

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java [hunk 7]
```diff
@@ -106,6 +107,8 @@ protected void createXmlDoc(Document doc) {
 			// }
 
 			constraints.appendChild(rule);
+
+			//createXmlConstraintDescription(doc, object.getConstraints().get(i), rule);
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:111]: Do not add code in comments. Remove?

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java [hunk 8]
```diff
@@ -140,6 +143,18 @@ protected void createXmlDoc(Document doc) {
...
 	}
+	
+	protected void createXmlConstraintDescription(Document doc, IConstraint constraint, Element parentRule) {
+		String desc = constraint.getDescription();
+		
+		final Element descr =
+				doc.createElement(DESCRIPTION);
+//			descr.setTextContent("\n"
+//				+ desc.replace("\r", "")
+//				+ "\n");
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:154]: as above

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java [hunk 9]
```diff
@@ -140,6 +143,18 @@ protected void createXmlDoc(Document doc) {
...
+		descr.setTextContent("Das ist ein Test");
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:155]: German

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 10]
```diff
@@ -472,6 +474,171 @@
...
+	public static final String REVERSE_FEATURE_ORDER =
+		"Reverse Feature Order";
+	public static final String SELECTION =
+		"Selection";
+	public static final String SHOW_HIDDEN_FEATURES =
+		"Show Hidden Features";
+	public static final String SHOW_COLLAPSED_CONSTRAINTS =
+		"Show Collapsed Constraints";
+	public static final String RENAMING_FEATURE =
+		"Renaming feature ";
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java:641]: Why are all these added here?

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ChangeConstraintDescriptionAction.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	public void run() {
+		String description =
+			"";
+		if (feature.getProperty().getDescription() != null) {
+			description =
+				feature.getProperty().getDescription();
+			description =
+				description.trim();
+		}
+		//TODO Team1 Story2 rename dialog
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ChangeConstraintDescriptionAction.java:53]: as above

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ChangeConstraintDescriptionAction.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+	@Override
+	public void run() {
+		String description =
+			"";
+		if (feature.getProperty().getDescription() != null) {
+			description =
+				feature.getProperty().getDescription();
+			description =
+				description.trim();
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ChangeConstraintDescriptionAction.java:51]: Why are you using the old formatter? Did you update to all recent commits from develop before formatting?

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateConstraintWithAction.java [hunk 13]
```diff
@@ -76,8 +77,9 @@ public void selectionChanged(SelectionChangedEvent event) {
 	 * @param featureName
 	 */
 	protected void updateConstraintActionText(String featureName) {
-		selectedFeature = featureName;
-		setText(CREATE_CONSTRAINT + (featureName.isEmpty() ? "" : " starting with \"" + featureName + "\""));
+		selectedFeature =	featureName;
+		setText(CREATE_CONSTRAINT	+ (featureName.isEmpty() ? ""	: " "	+ STARTING_WITH	+ " \""	+ featureName	+ "\""));
+
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/CreateConstraintWithAction.java:82]: Looks like the formatter was not applied here.

File: tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/TAbstractFeatureModelReaderWriter.java [hunk 14]
```diff
@@ -245,6 +245,7 @@ public void testDescription() {
 
 	private final IFeatureModel writeAndReadModel() throws UnsupportedModelException {
 		IFeatureModel newFm = null;
+
```
Comment (tthuem) [inline tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/TAbstractFeatureModelReaderWriter.java:248]: Why changed?

File: tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/xml/TXmlFeatureModelFormat.java [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+			factory.createConstraint(newFm, n1);
+		c1.setDescription("Test Write Description 1");
+
+		final IConstraint c2 =
+			factory.createConstraint(newFm, n2);
+		c2.setDescription("Test Write Description 2");
+		// IConstraint c3 = factory.createConstraint(newFm, n3);
+		newFm.addConstraint(c1);
+		newFm.addConstraint(c2);
+		// newFm.addConstraint(c3);
```
Comment (tthuem) [inline tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/xml/TXmlFeatureModelFormat.java:198]: Numerous lines with code in comments here.

File: tests/de.ovgu.featureide.fm.core-test/src/testFeatureModels/.featureide/constrainDescriptionWriteTwoRulesTest.xml.xml/constrainDescriptionWriteTwoRulesTest.xml.xml.xml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (tthuem) [inline tests/de.ovgu.featureide.fm.core-test/src/testFeatureModels/.featureide/constrainDescriptionWriteTwoRulesTest.xml.xml/constrainDescriptionWriteTwoRulesTest.xml.xml.xml:1]: Why added something from .featureide folder? Should not be necessary, right?

File: tests/de.ovgu.featureide.fm.core-test/src/testFeatureModels/constraintDescriptionTwoRulesTest.xml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+				</conj>
+			</rule>
+			<rule>
+			
+				<imp>
+					<var>A</var>
+					<var>B</var>
+				</imp>
+			<description>
+				Test Description
```
Comment (tthuem) [inline tests/de.ovgu.featureide.fm.core-test/src/testFeatureModels/constraintDescriptionTwoRulesTest.xml:27]: I would use different descriptions to alse check that descriptions are not messed up. And maybe also a third rule without any description.

**Review by tthuem — CHANGES_REQUESTED** [review]
I had a look already, but seems there are numerous changes necessary. Please revise and start new review request.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java (hunk 1 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java:30]: Yes, we are using it in the setDescription Method to check that the description parameter is not null.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java (hunk 2 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:65]: okay, removed.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java (hunk 3 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:303]: removed all

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java (hunk 4 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:421]: Checked the meaning of the line and removed it after realizing that it makes no sense there.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java (hunk 6 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:490]: In this line the DESCRIPTION is supposed to be ignored. So the method should return without adding any nodes, just return the current list. The actual readout of the description happens at a different point.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java [hunk 18]
```diff
@@ -190,5 +194,18 @@ public void setNode(Node node) {
 	public String toString() {
 		return "AConstraint [propNode=" + propNode + "]";
 	}
+	
+	public void setDescription(@Nonnull final CharSequence description) {
+		this.description =
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/AConstraint.java:199]: Format

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 19]
```diff
@@ -455,9 +455,11 @@
 	public static final String CALCULATE_TAUTOLOGY_CONSTRAINTS = "Calculate Tautology Constraints";
 	public static final String CHANGE_DESCRIPTION = "Change Description";
 	public static final String FEATURE_DESCRIPTION = "Feature Description";
+  public static final String CONSTRAINT_DESCRIPTION = "Constraint Description";
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java:458]: Indent line 458 and 462

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 20]
```diff
@@ -1001,6 +1168,7 @@
 	public static final String INVALID_COMMA_ = "invalid, ";
 	public static final String DOES_NOT_EXIST_ = " does not exist.";
 	public static final String PLEASE_ENTER_A_DESCRIPTION_FOR_FEATURE_ = "Please enter a description for feature '";
+  public static final String PLEASE_ENTER_A_DESCRIPTION_FOR_CONSTRAINT_ = "Please enter a description for constraint '";
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java:1171]: As above

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java [hunk 21]
```diff
@@ -498,6 +500,7 @@ private void createActions() {
 		expandAllAction.setImageDescriptor(FmOutlinePageContextMenu.IMG_EXPAND); // icon for expand added
 		abstractAction = new AbstractAction(this, featureModel, (ObjectUndoContext) featureModel.getUndoContext());
 		changeFeatureDescriptionAction = new ChangeFeatureDescriptionAction(this, featureModel, null);
+    changeConstraintDescriptionAction =	new ChangeConstraintDescriptionAction(this, featureModel, null);
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java:503]: As above

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java [hunk 22]
```diff
@@ -655,46 +658,56 @@ private void fillContextMenu(IMenuManager menu) {
...
+
 			menu.add(new Separator());
 			menu.add(calculateDependencyAction);
-			menu.add(reverseOrderAction);
-			menu.add(legendAction);
 			menu.add(new Separator());
-		} else if (editConstraintAction.isEnabled() && !connectionSelected) {
+
+      // if a constraint is selected
+		} else if (editConstraintAction.isEnabled()	&& !connectionSelected) {
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureDiagramEditor.java:688]: Remove the tab stop before &&

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/DeleteAllAction.java [hunk 23]
```diff
@@ -74,8 +74,7 @@ public void run() {
 	 */
 	@Override
 	protected void updateProperties() {
-		setEnabled(!feature.getStructure().isRoot() && feature.getStructure().hasChildren());
-		setChecked(false);
+		setEnabled(!feature.getStructure().isRoot()	&& feature.getStructure().hasChildren());
```
Comment (Henningson) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/DeleteAllAction.java:77]: As above

File: tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/xml/TXmlFeatureModelFormat.java [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+		}
+
+		assertEquals(gFM.getVisibleFeatures().size(), gfmCollapsed.getVisibleFeatures().size() + notVisible);
+
+		assertEquals(gFM.getVisibleFeatures().size(), gfmNotCollapsed.getVisibleFeatures().size());
+	}
+
+	@Test
+	public void testConstraintDescription() throws FileNotFoundException, UnsupportedModelException {
+		String constraindescriptionFromXml =
```
Comment (Henningson) [inline tests/de.ovgu.featureide.fm.core-test/src/de/ovgu/featureide/fm/core/io/xml/TXmlFeatureModelFormat.java:110]: Change constraindescriptionFromXml to constraintDescriptionFromXml

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java (hunk 7 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:111]: removed

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java (hunk 8 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:154]: removed

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java (hunk 9 above)
Comment (DawidSA2017) [reply plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelWriter.java:155]: removed

**Pull request merged**